clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java

```diff
@@ -0,0 +1,864 @@
+     */
+    private void balance(Map<String, List<TopicPartition>> currentAssignment,
+                         Map<TopicPartition, ConsumerGenerationPair> prevAssignment,
+                         List<TopicPartition> sortedPartitions,
+                         List<TopicPartition> unassignedPartitions,
+                         TreeSet<String> sortedCurrentSubscriptions,
+                         Map<String, List<TopicPartition>> consumer2AllPotentialPartitions,
+                         Map<TopicPartition, List<String>> partition2AllPotentialConsumers,
+                         Map<TopicPartition, String> currentPartitionConsumer,
+                         boolean revocationRequired) {
+        boolean initializing = currentAssignment.get(sortedCurrentSubscriptions.last()).isEmpty();
+        boolean reassignmentPerformed = false;
+
+        // assign all unassigned partitions
+        for (TopicPartition partition: unassignedPartitions) {
+            // skip if there is no potential consumer for the partition
+            if (partition2AllPotentialConsumers.get(partition).isEmpty())
+                continue;
+
+            assignPartition(partition, sortedCurrentSubscriptions, currentAssignment,
+                consumer2AllPotentialPartitions, currentPartitionConsumer);
+        }
+
+        // narrow down the reassignment scope to only those partitions that can actually be reassigned
+        Set<TopicPartition> fixedPartitions = new HashSet<>();
+        for (TopicPartition partition: partition2AllPotentialConsumers.keySet())
+            if (!canParticipateInReassignment(partition, partition2AllPotentialConsumers))
+                fixedPartitions.add(partition);
+        sortedPartitions.removeAll(fixedPartitions);
+        unassignedPartitions.removeAll(fixedPartitions);
+
+        // narrow down the reassignment scope to only those consumers that are subject to reassignment
+        Map<String, List<TopicPartition>> fixedAssignments = new HashMap<>();
+        for (String consumer: consumer2AllPotentialPartitions.keySet())
+            if (!canParticipateInReassignment(consumer, currentAssignment,
+                consumer2AllPotentialPartitions, partition2AllPotentialConsumers)) {
+                sortedCurrentSubscriptions.remove(consumer);
+                fixedAssignments.put(consumer, currentAssignment.remove(consumer));
+            }
+
+        // create a deep copy of the current assignment so we can revert to it if we do not get a more balanced assignment later
+        Map<String, List<TopicPartition>> preBalanceAssignment = deepCopy(currentAssignment);
+        Map<TopicPartition, String> preBalancePartitionConsumers = new HashMap<>(currentPartitionConsumer);
+
+        // if we don't already need to revoke something due to subscription changes, first try to balance by only moving newly added partitions
+        if (!revocationRequired) {
```

Comment:
This is the one change to the sticky assignment algorithm -- in some rare cases the assignment logic will sacrifice stickiness for balance unnecessarily, triggering a third rebalance under cooperative mode (see [KAFKA-8767](https://issues.apache.org/jira/browse/KAFKA-8767) for more context).
This doesn't fix this entirely, but did reduce the occurrences by a factor of more than 2. I feel this is fine for now as the situation should be rare (mostly random subscription changes) and not too costly, and we can further optimize later.

Reply:
Changing subscriptions in a random fashion should be rare, but I'm more curious of why the algorithm would fail to find the optimal solution, and if it is could be triggered by other even more common situations?

By reading the description of the sticky assignor from KIP-441, I think the root cause may be in the iterative balancing phase:

```
When considering the best move for a partition, it first checks to see if that partition is currently hosted on a consumer that is unbalanced with respect to the prior host of that partition. In this case, it just moves the partition back to the prior host. This is essentially a short-circuit for the case where a partition has become "unstuck" and restoring stickiness could actually improve balance. If we get past that short-circuit, then we just propose to move the partition to the consumer that can host it and has the smallest current assignment.
```

Such random walks algorithm can only achieve sub-optimal solution indeed; maybe we can change this behavior to consider "switching two partitions so that the total balance of the cluster does not change, but we can achieve better stickiness" as a valid move before considering for convergence? Just a thought.

Reply:
Yeah, I didn't see substantial improvement trying that but perhaps there is a more clever way to do so. I opened a ticket to look into this further but consider it an optimization as future work. Personally, I feel it's an interesting problem -- but doesn't merit blocking progress on 429.

Does that sound reasonable? I also think most use cases will not hit this -- and if they do, it's not incorrect just annoying :)

Reply:
Yeah as long as you feel that no other more common situations would hit this, then it should not block KIP-429.